File: socket_server/src/utils/socket/message_io.ts
import {Namespace} from 'socket.io';

export function messageSocket(messageIO: Namespace, recentIO: Namespace) {
  messageIO.on('connection', (socket) => {
    const q = socket.handshake.query.chatRoomId;
    if (!q) return;
    const chatRoomId = q as string;
    socket.join(chatRoomId);
    console.log('Chat In', chatRoomId);

    socket.on('new_message', (msg) => {
      messageIO.to(chatRoomId).emit('new_message', msg);
    });

    socket.on('read', (ids) => {
      messageIO.to(chatRoomId).emit('read', ids);
    });

    socket.on('update_recent', (data) => {
      const roomIds = data['userIds'];
      /// socket.io V4
      recentIO.to(roomIds).emit('update', data);
    });

    socket.on('disconnect', () => {
      socket.leave(chatRoomId);
      console.log('Chat Disconnrect');
    });
  });
}

File: socket_flutter/lib/src/socket/message_io.dart
import 'package:socket_flutter/src/model/message.dart';
import 'package:socket_flutter/src/socket/socket_base.dart';

import '../screen/main_tab/message/message_controller.dart';
import '../screen/main_tab/message/message_extention.dart';

class MessageIO extends SocketBase {
  @override
  NameSpace get nameSpace => NameSpace.message;
  @override
  Map<String, dynamic> get query => {"chatRoomId": chatRoomId};

  final MessageController controller;
  MessageIO(this.controller);

  MessageExtention get extention => controller.extention;
  String get chatRoomId => extention.chatRoomId;

  //受信
  void addNewMessageListner() {
    socket.on("new_message", (msg) {
      final newMessage = Message.fromMap(msg);

      if (!newMessage.isCurrent) {
        extention.updateRead(message: newMessage);
        sendUpdateRead([newMessage.id]);
      }

      controller.messages.insert(0, newMessage);
    });
  }

  void addReadListner() {
    socket.on("read", (value) {
      final uid = value["uid"];
      final ids = value["ids"];
      final List<String> temp = List.castFrom(ids);
      temp.forEach((String id) => controller.readUI(id, uid));
    });
  }

  /// 送信
  void sendNewMessage(Message message) {
    socket.emit("new_message", message.toMap());
  }

  void sendUpdateRead(List<String> messageIds) {
    final data = {"uid": extention.currentUser.id, "ids": messageIds};
    socket.emit("read", data);
  }

  void sendUpdateRecent(List<String> userIds) {
    final Map<String, dynamic> data = {
      "chatRoomId": chatRoomId,
      "userIds": userIds,
    };

    socket.emit("update_recent", data);
  }
}
